File: 219 Contains Duplicate II/containsNearbyDuplicate.cpp
//  Time complexity: O(n)
// Space complexity: O(k)

class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        
        unordered_set<int> record;
        for(int i = 0; i < nums.size(); i++){
            
            if( record.find(nums[i]) != record.end() )
                return true;
            
            record.insert(nums[i]);
            
            // Keep the number of elements in record <= k, using record to keep the size of sliding window to k+1
            if( record.size() == k+1)
                record.erase(nums[i-k]);
        }
        return false;
    }
};

File: 219 Contains Duplicate II/containsNearbyDuplicate.py
class Solution(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        record = set()
        for i,element in enumerate(nums):
            
            if element in record:
                return True
            record.add(element)
            if len(record) == k+1:
                record.remove(nums[i-k])
        
        return False